pvlib/mlfm.py

```diff
@@ -0,0 +1,804 @@
+
+    'pr_dc':        'black',       # 000 000 000
+}
+
+
+def mlfm_meas_to_norm(dmeas, ref):
+    '''
+    Convert measured power, current and voltage to normalized values.
+
+    Parameters
+    ----------
+    dmeas : DataFrame
+        Measurements. Must include columns:
+
+        * `'poa_global'` global plane of array irradiance [W/m^2]
+        * `'temp_module'` module temperature [C]
+        * `'p_mp'` - power at maximum power point [W]
+
+        May include optional columns:
+
+        * `'i_sc'` - current at short circuit condition [A]
+        * `'v_oc'` - voltage at open circuit condition [V]
+        * `'i_mp'` - current at maximum power point [A]. Must be accompanied
+          by `'i_sc'`.
+        * `'v_mp'` - voltage at maximum power point [V]. Must be accompanied
+          by `'v_oc'`.
+        * `'r_sc'` - inverse of slope of IV curve at short circuit condition.
+          Requires both `'i_sc'` and `'v_oc'`. [Ohm]
+        * `'r_oc'` - inverse slope of IV curve at open circuit condition.
+          Requires both `'i_sc'` and `'v_oc'` [Ohm]
+
+    ref : dict
+        Reference values. Must include:
+
+        * `'p_mp'` - Power at maximum power point at Standard Test Condition
+          (STC). [W]
+        * `'gamma_pdc'` - Temperature coefficient of power at STC. [1/C]
+
+        May include:
+
+        * `'i_sc'` - Current at short circuit at STC. Required if `'i_sc'` is
+          present in ``dmeas``. [A]
+        * `'v_oc'` - Voltage at open circuit at STC. Required if `'v_oc'` is
+          present in ``dmeas``. [A]
+        * `'beta_v_oc'` - Temperature coefficient of open circuit voltage at
+          STC. Required if `'v_oc'` is present in ``dmeas``. [1/C]
+
+    Returns
+    -------
+    dnorm : DataFrame
+        Normalised values.
+
+        * `'pr_dc'` is `'p_mp'` normalised (divided) by reference `'p_mp'` \
+            and by `'poa_global'` in kW/m^2.
+        * `'pr_dc_temp_corr'` is `'pr_dc'` adjusted to 25C.
+        * Columns `'i_sc'`, `'i_mp'`, `'v_oc'`, `'v_mp'`, `'v_oc_temp_corr'`,
+          `'r_sc'`, `'r_oc'`, `'i_ff'`, `'v_ff'` are returned when the
+          the corresponding optional columns are included in ``dmeas``.
+
+    References
+    ----------
+    .. [1] Steve Ransome (SRCL) and Juergen Sutterlueti (Gantner Instruments)
+       "Quantifying Long Term PV Performance and Degradation under Real Outdoor
+       and IEC 61853 Test Conditions Using High Quality Module IV Measurements"
+       36th EU PVSEC, Marseille, France. September 2019
+    '''
+    dnorm = pd.DataFrame()
+
+    dnorm['pr_dc'] = dmeas['p_mp'] / ref['p_mp'] \
+        / (dmeas['poa_global'] / G_STC)
+
+    # temperature corrected
+    dnorm['pr_dc_temp_corr'] = (
+        dnorm['pr_dc'] *
+        (1 - ref['gamma_pdc']*(dmeas['temp_module'] - T_STC)))
+
+    if 'i_sc' in dmeas.columns:
+        dnorm['i_sc'] = dmeas['i_sc'] / (dmeas['poa_global'] / G_STC) \
+            / ref['i_sc']
```

Comment:
Nit: This code would read a bit quicker if it followed the same arrangement as for `dnorm['pr_dc']`, i.e.,

```python
dnorm['i_sc'] = dmeas['i_sc']  / ref['i_sc'] / \
    (dmeas['poa_global'] / G_STC)
```

Reply:
>     # calculate normalised loss coefficients
>     if 'i_sc' in dmeas.columns:
>         dnorm['i_sc'] = (dmeas['i_sc'] / ref['i_sc']
>                          / (dmeas['poa_global'] / G_STC))